File: 2016/01.cpp
#include <fmt/core.h>
#include <set>

#include "general.hpp"

using namespace std;
using namespace general;

int main() {
    string data = get_input("01.data");
    const auto lines = split(data, "\n");

    pair<int, int> pos{0, 0};
    pair<int, int> dir{0, 1};

    set<pair<int, int>> positions{};
    positions.insert(pos);

    int first_double = -1;

    for (auto s : split(*lines.begin(), ", ")) {
        int c;
        // string_view to int - no errors possible (for aoc specifically)
        (void)from_chars(s.begin()+1, s.end(), c);
        dir = s[0] == 'R' ? make_pair(get<1>(dir), -get<0>(dir)) : make_pair(-get<1>(dir), get<0>(dir));

        for (int i = 1; i <= c && first_double < 0; i++) {
            pair<int, int> tmp = {get<0>(pos)+i*get<0>(dir), get<1>(pos)+i*get<1>(dir)};
            if (first_double < 0 && positions.contains(tmp)) {
                first_double = abs(get<0>(tmp))+abs(get<1>(tmp));
                break;
            }
            positions.insert(tmp);
        }
        pos = {get<0>(pos)+c*get<0>(dir), get<1>(pos)+c*get<1>(dir)};
    }
    fmt::print("{}\n", abs(get<0>(pos))+abs(get<1>(pos)));
    fmt::print("{}\n", first_double);
    return 0;
}

// year 2016
// solution for 01.01: 300
// solution for 01.02: 159

File: 2016/01.py
#!/usr/bin/env python3

import sys
sys.path.append('../General')
from utility import *

def main():

    lines = open_data("01.data")

    pos = (0,0)
    d = (0,-1)
    all_pos = set()
    part_2 = None
    for l in lines[0].split(", "):
        d = rotate(d, l[0] != "L")
        old_pos = pos
        pos = add(pos, mul(d, int(l[1:])))

        if part_2 is None:
            for i in range(1, int(l[1:])):
                new_pos = (old_pos[0]+d[0]*i, old_pos[1]+d[1]*i)
                if new_pos in all_pos:
                    part_2 = new_pos
                all_pos.add(new_pos)

    print(sum(map(abs, pos)))
    print(sum(map(abs, part_2)))


if __name__ == "__main__":
    main()

# year 2016
# solution for 01.01: 300
# solution for 01.02: 159
